subscriber_notifee.go

```diff
@@ -0,0 +1,147 @@
+package dht
+
+import (
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	"github.com/libp2p/go-eventbus"
+
+	ma "github.com/multiformats/go-multiaddr"
+
+	"github.com/jbenet/goprocess"
+)
+
+// subscriberNotifee implements network.Notifee and also manages the subscriber to the event bus. We consume peer
+// identification events to trigger inclusion in the routing table, and we consume Disconnected events to eject peers
+// from it.
+type subscriberNotifee IpfsDHT
+
+func (nn *subscriberNotifee) DHT() *IpfsDHT {
+	return (*IpfsDHT)(nn)
+}
+
+func (nn *subscriberNotifee) subscribe(proc goprocess.Process) {
+	dht := nn.DHT()
+
+	dht.host.Network().Notify(nn)
+	defer dht.host.Network().StopNotify(nn)
+
+	var err error
+	evts := []interface{}{
+		&event.EvtPeerIdentificationCompleted{},
+	}
+
+	// subscribe to the EvtPeerIdentificationCompleted event which notifies us every time a peer successfully completes identification
+	sub, err := dht.host.EventBus().Subscribe(evts, eventbus.BufSize(256))
+	if err != nil {
+		logger.Errorf("dht not subscribed to peer identification events; things will fail; err: %s", err)
+	}
+	defer sub.Close()
+
+	dht.plk.Lock()
+	for _, p := range dht.host.Peerstore().Peers() {
```

Comment:
`dht.host.Network().Peers()`? Those'll always be connected.

Reply:
Note: we could have a lot of peers in the peerstore (if persisted).

Reply:
Ah good idea, thanks